### ISSUE TITLE
bring forward nice login prompts from container linux

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-tracker#36]
When you log in to container linux you get a nice prompt that tells you some information about the machine. 

```
$ ssh core@192.168.122.136
Container Linux by CoreOS stable (1800.7.0)
Failed Units: 1
  example.service
core@localhost ~ $ 
```

On the serial console there is a nice message telling you what the IP address is for the machine.
```
This is localhost (Linux x86_64 4.14.16-coreos) 14:19:32
SSH host key: SHA256:yP+/44/bfuj6UKHdUwAVURsO3y6haKLKfSFNcnmn7bY (ECDSA)
SSH host key: SHA256:gGDZ/JQzwL76UpT29dyZ/M6Zua7QvGyegP8aTLc/D+Y (DSA)
SSH host key: SHA256:nQEysCYP3hZgkus2+e28KQGrs0pRI2NOgJGQ6L8PnyU (RSA)
SSH host key: SHA256:A3c6toZ3/eTMKNDmyyG9CYUSWsdSunmTeOC68iuDfAg (ED25519)
eth0: 192.168.122.36 fe80::5054:ff:fe85:43a6
```

we should carry these nice features forward for Fedora CoreOS.

**Comment by cgwalters** [comment]
See also https://lists.fedoraproject.org/archives/list/server@lists.fedoraproject.org/thread/55SWJA6ERCC3ZPAGEQFJK5EGAXTFX7OJ/

**Comment by cgwalters** [comment]
Also, one thing I think is very important is to report whether or not Ignition ran.

I'd also note if any ssh keys or users were added.

**Comment by lucab** [comment]
@dustymabe for reference, the first comes from https://github.com/coreos/init/blob/master/scripts/motdgen and the second from https://github.com/coreos/init/blob/master/scripts/issuegen.

**Comment by lucab** [comment]
> Also, one thing I think is very important is to report whether or not Ignition ran.

@cgwalters what are you specifically looking for here? I'm asking because we can be sensibly report whether ignition ran *on the current boot*.
But for the generic case (any boot) the answer is always "yes": if you are in the rootfs, then ignition ran at some arbitrary point in the past, otherwise machine provisioning would have failed in initramfs.

**Comment by cgwalters** [comment]
> if you are in the rootfs, then ignition ran at some arbitrary point in the past, otherwise machine provisioning would have failed in initramfs.

True, though we've had bugs where it hasn't run.  There's also the case of "I want to perform a system reset", or "I'm testing something and want Ignition to re-run".

**Comment by rfairley** [comment]
Off-topic from the login prompts, but while ignition is in discussion: do we have an "official" way to re-run ignition? So far `touch /boot/coreos-firstboot` will re-run it. But I found more needs to be supplied for it to behave the same as if it were a first boot, e.g. a `coreos.config.url` to get it to re-fetch the config data.

Would we look to have some factory reset button available? It might take the form of a shell script, and would set things up so the next reboot is like a first boot. Its use case would be directed at testing as @cgwalters mentioned. Or is this already achieved by `touch /boot/coreos-firstboot`?

**Comment by ajeddeloh** [comment]
@rfairley As of right now there isn't a way to do a "factory reset" since we can't redeploy the ostree from the initramfs. It's something that I think would be possible afterwards though. Basically just use Ignition to recreate the filesystems (including var) and verify the partitions. redeploy, rerun the files stage and blammo, you're factory reset.

**Comment by cgwalters** [comment]
Why do we need to redeploy?  Since everything except `/etc` and `/var` is immutable, factory reset is basically:

```
rsync -av --delete /usr/etc/ /etc/
rm -rf /var/*
touch /boot/coreos-firstboot
```

right?  

(Except of course today you want to keep at least `/etc/fstab`)

**Comment by ajeddeloh** [comment]
Right, but having Ignition do it makes it simpler/less error prone. No weirdness of having the running system using the `/etc` it's actively reseting. No logs trying to be written to places in `/var` that don't exist. No weirdness with `/etc/fstab`. No special handling of extra partitions you brought with you. Plus you can keep the node as is until you go down for reboot.

**Comment by cgwalters** [comment]
> Right, but having Ignition do it makes it simpler/less error prone. 

Yep, agreed!  

(Though most of this is really "in the initramfs" in general - if we still have the initramfs filesystem around we could also pivot to it at shutdown time and do system reset there, but eh, might as well do it on boot)

**Comment by rfairley** [comment]
Thinking about how the `motd` could look in RHCOS/FCOS, I'm thinking the following features should be included:
- have PAM point to the `/etc/motd` and `/etc/mod.d` locations, from a config in `/etc/pam.d` (which I think we control?)
- dynamically generate a motd with updated information using a systemd unit (like what `motdgen` does), each boot
- make it easy for users to add in additional motd messages, by dropping a file in some `motd.d` directory (this might be helpful in FCOS?)

**motd file**

[This](https://github.com/rfairley/fedora-coreos-login-messages/blob/master/etc/pam.d/sshd) is the `sshd` PAM configuration in Fedora 29 Beta 1.1.
- One way to let PAM know to look in `/run` for the `motd` is pass an argument to the `pam_motd` module in `/etc/pam.d/sshd` to look at `/run/*-coreos/motd` instead of `/etc/motd` (can do the same for the directory `motd.d`).
- Another option is to use a symlink like on CL currently (version `1897.0.0+2018-09-12-2016`): `/etc/motd -> /run/coreos/motd`. Possibly, we have PAM search `/etc/motd` and `/etc/motd.d` (the default config) and follow symlinks to `/run/*-coreos/motd` and `/run/*-coreos/motd.d`. This would work nicely for authentication that does not go through PAM but looks at `/etc/issue`.

**issue file**

As for the `issue`, as I understand `agetty` will look in `/etc/issue` and `/etc/issue.d` to generate the issue displayed before login on a console. The issue file location can also be specified in PAM by the module `pam_issue`, but from the man page it does not look like an issue directory can be specified. If all we need is the `issue` file, then specifying the issue location through PAM would work nicely. Need to test this - to make sure something like "PAM prints the issue specified using `pam_issue`, then agetty finds it and prints it again" doesn't happen.

If we were to include an `issue.d` where users could drop in additional issue messages, I think it would be fine to symlink `/etc/issue.d` to somewhere like `/run/issue.d` (`issuegen` would have to generate `/run/issue` with files from somewhere else like `/run/coreos/issue.d` in that case).

**Comment by lucab** [comment]
@rfairley this generally looks fine, I'll gladly have a further look once you get all the bits settled.

A couple of additional feedback points below.

> from a config in /etc/pam.d (which I think we control?)

We control its vendor-equivalent path at `/usr/lib/pam.d/`. It can be done that way, yes.

> motd file: one way to let PAM know ... Another option is to use a symlink ...

The difference between the two is that to customize motd (let's say to a static hardcoded content) with the former one needs to tweak PAM configuration, while with the latter one can simply blow the symlink and place a file under the usual /etc location. I somehow prefer the former and simply avoiding vendor files in /etc, but other people may disagree.
(As a wishlist, it would be nice to augment `pam_motd` so that it takes a list of files to try in order, so that we can specify `motd=/etc/motd:/run/motd:/usr/lib/motd` or multiple `motd=`, and make the whole thing simpler).

> issuegen would have to generate /run/issue with files from somewhere else like /run/coreos/issue.d in that case

Ack on that, it would be better to scope it under `/run/coreos` or `/run/coreos-issuegen` now.
On CL, I think we don't use `pam_issue`, just plain agetty. In both cases, I have the same wishlist of being able to fallback through etc-run-usrlib looking for an `issue` file.

**Comment by rfairley** [comment]
> avoiding vendor files in /etc

If avoiding vendor files in `/etc`, would this mean we'd intend for the user to add a `/etc/motd` if they'd like to customize the motd, and if that file doesn't exist, fall back to the one we generate in e.g. `/run/motd`?

I do like that idea of trying different motd locations to see if the file exists and using only the first one that exists - makes it a lot easier for the user to override our generated one in `/run/motd`. `/usr/lib/motd` could be a static one which acts as a fallback if the one in `/run/motd` somehow does not get generated.

May see if the change can be implemented upstream easily - would be very nice to have the logic of trying different locations, from one line in the PAM config. Otherwise, that logic could be implemented by a systemd unit as well - e.g. to copy the first `motd` file that exists to some single location that PAM will read from. Will start with the systemd units, to get a first iteration working.

**Comment by rfairley** [comment]
Added a .specfile to package the scripts/units/config for this as an RPM [here](https://github.com/rfairley/fedora-coreos-login-messages/blob/master/coreos-ux.spec). There are a few things left to sort out - we can discuss tomorrow/Friday (see [this](https://github.com/rfairley/fedora-coreos-login-messages#issues-to-figure-out)).

One question is how should this rpm be named? For now I named it `coreos-ux`. Will we later combine the ux-related files (motd*, issue*, profile*) with other "base" systemd units/scripts/configs that might also be shipped on *COS (and not necessarily relate to UX)? Thinking of https://github.com/coreos/baselayout.

**Comment by dustymabe** [comment]
Nice work @rfairley - is there a built RPM somewhere I can pick up and test out?

One thing I'd like to consider here is making this rpm generic enough to be installed on non CoreOS systems and be able to deliver the same great benefits there. How much of the work you have done is CoreOS specific?

**Comment by rfairley** [comment]
@dustymabe thanks for the feedback! Here is a scratch build on Koji: https://koji.fedoraproject.org/koji/taskinfo?taskID=29921437. Tested this just now on `fedora/28-cloud-base`, but fedora 28 or RHCOS should also work. Right now there are 4 rpms - split some things into subpackages.

```
# curl --remote-name-all https://kojipkgs.fedoraproject.org//work/tasks/1437/29921437/coreos-ux-0.1-1.fc28.noarch.rpm https://kojipkgs.fedoraproject.org//work/tasks/1437/29921437/coreos-ux-issuegen-0.1-1.fc28.noarch.rpm https://kojipkgs.fedoraproject.org//work/tasks/1437/29921437/coreos-ux-motdgen-0.1-1.fc28.noarch.rpm https://kojipkgs.fedoraproject.org//work/tasks/1437/29921437/coreos-ux-profile-0.1-1.fc28.noarch.rpm
# dnf install coreos-ux-*
```

> How much of the work you have done is CoreOS specific?

**Comment by rfairley** [comment]
> How much of the work you have done is CoreOS specific?

Right now just the naming and the `coreos` scoping of the directories (e.g. `/usr/lib/coreos/...`) are CoreOS-specific - these could easily be renamed to something general. The data in the generated `motd` and `issue` are sourced in a general way as well, e.g. from `/usr/lib/os-release` - this could apply to other distros as well.

**Comment by dustymabe** [comment]
> Right now just the naming and the `coreos` scoping of the directories (e.g. `/usr/lib/coreos/...`) are CoreOS-specific - these could easily be renamed to something general.

:+1: 



> The data in the generated `motd` and `issue` are sourced in a general way as well, e.g. from `/usr/lib/os-release` - this could apply to other distros as well.

That's awesome! If it works with other distros that would be great, but we don't necessarily need to bend over backwards to make that happen. I think working in the rest of fedora would be a big +1 though because we get more people using it and testing it and contributing back too.

**Comment by dustymabe** [comment]
> @dustymabe thanks for the feedback! Here is a scratch build on Koji

woot. just grabbed this and it looks pretty good:

```
[dustymabe@media ~]$ ssh core@192.168.122.9
Fedora CoreOS (preview)
---

test: Failed systemd units will show here, if there are any.
Failed Units: 1
  example.service
```

Some comments:

- do we need the `test: Failed systemd units will show here, if there are any.` line?
- I don't see the `eth0: x.x.x.x` line on the serial console of the machine when I boot it up as illustrated in [the original comment](https://github.com/coreos/fedora-coreos-tracker/issues/36#issue-354763917). Is that expected?

**Comment by rfairley** [comment]
+1, agreed on having this working in Fedora.

Thank you for trying it out!

> do we need the test: Failed systemd units

We do not need this line - will take this out. This was just to confirm before that the `/etc/profile.d/coreos-profile.sh` script would print something (I should have tested it with a systemd unit known to fail - will do that next time).

**Comment by rfairley** [comment]
> I don't see the eth0: x.x.x.x line on the serial console of the machine when I boot it up as illustrated in the original comment. Is that expected?

The package is behaving as expected right now. It may be printing an `/etc/issue` that already existed - the symlink to `/run/issue` does not get made if a file already exists there. If the symlink `/etc/issue -> /run/issue` is present, then `systemctl start issuegen.path issuegen.service motdgen.path motdgen.service` or rebooting should generate the issue with the `eth0: x.x.x.x` line.

 I actually did just notice across a bug now, as rebooting was not showing the `eth0: x.x.x.x` line. `issuegen` was checking if `/run/motd.d` existed instead of `/run/issue.d`. Fixing that!

**Comment by rfairley** [comment]
@dustymabe another build: https://koji.fedoraproject.org/koji/taskinfo?taskID=29922963

```
# curl --remote-name-all https://kojipkgs.fedoraproject.org//work/tasks/2963/29922963/coreos-ux-0.1-1.fc28.noarch.rpm https://kojipkgs.fedoraproject.org//work/tasks/2963/29922963/coreos-ux-issuegen-0.1-1.fc28.noarch.rpm https://kojipkgs.fedoraproject.org//work/tasks/2963/29922963/coreos-ux-motdgen-0.1-1.fc28.noarch.rpm https://kojipkgs.fedoraproject.org//work/tasks/2963/29922963/coreos-ux-profile-0.1-1.fc28.noarch.rpm
# dnf install coreos-ux-*
# systemctl reboot
# cat /run/issue
This is \n (\s \m \r) \t
eth0: \4{eth0} \6{eth0}
```

This should now be the correct behaviour. If the symlink `/etc/issue -> /run/issue` is created, then you should see the eth0 line in the serial console.

**Comment by dustymabe** [comment]
+1 will test that in the next hour

**Comment by rfairley** [comment]
+1 thanks. Also, am adding in showing the ssh keys just now. Will post an updated build with that included once done

**Comment by rfairley** [comment]
Added showing the ssh keys in. Will keep the installation steps posted here: https://github.com/rfairley/fedora-coreos-login-messages#installation - these can now be tried and should show eth0 and ssh host keys in `issue`.

**Comment by rfairley** [comment]
Can be tested by following: https://github.com/rfairley/fedora-coreos-login-messages#installation. This should show the issue file as expected after completing step 7.

One remaining problem is that the file symlinked as `/etc/motd.d/fedora-user-messages.motd -> /run/fedora-user-messages.motd` is not being displayed by PAM after SSH login. `cat /run/fedora-user-messages.motd` shows the file is being generated correctly by the systemd units. Copying the generated file so it is a regular file in`/etc/motd.d/` causes the motd to be shown after SSH login - this makes me think symlinks in `/etc/motd.d/` are not being followed by PAM. Looking into this to see if cockpit got this to work.

**Comment by dustymabe** [comment]
looking good! on serial console:

```
Fedora 29 (CoreOS)
Kernel 4.18.5-300.fc29.x86_64 on an x86_64 (ttyS0)

----
This is localhost.localdomain (Linux x86_64 4.18.5-300.fc29.x86_64) 19:10:11
----
eth0: 192.168.122.159 fe80::5054:ff:fe8c:611f
localhost login:
```

on ssh login:

```
$ ssh core@192.168.122.158
The authenticity of host '192.168.122.158 (192.168.122.158)' can't be established.
ECDSA key fingerprint is SHA256:deyAg2C+sofoi69cLEeMe+X4WIuE3r66otGqGu7Vu0k.
ECDSA key fingerprint is MD5:af:3f:73:38:b6:88:d6:26:cf:a8:ad:2e:dd:aa:fd:9f.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '192.168.122.158' (ECDSA) to the list of known hosts.
Fedora CoreOS (preview)
---

Failed Units: 1
  example.service

```

right now the rpms are named `fedora-user-messages-*`. Which we might have to rebrand if this package ever got passed along to RHEL/centos. Should we just name it something generic? @sgallagher do you have a suggestion?

**Comment by rfairley** [comment]
Thank you for trying it @dustymabe !

Just to add on the motd, the motd will show the following including some information from `rpm-ostree`:
```
# cat /etc/motd.d/fedora-user-messages.motd 
Fedora (28 (Atomic Host))

---- available users
  root
  vagrant

---- rpm-ostree
Update available:
  111 packages to upgrade
  1 packages to remove
  4 new packages
```
This takes information from `rpm-ostree upgrade --check` (generated at https://github.com/rfairley/fedora-coreos-login-messages/blob/master/usr/lib/coreos/motdgen#L68). Any thoughts on this - too verbose? 

Right now, the above motd is not getting displayed on SSH login, it is symlinked at `/etc/motd.d/fedora-user-messages.motd` to `/run/fedora-user-messages.motd` - checking what needs to be done to have PAM display this. Any ideas on things to try are appreciated!

**Comment by LorbusChris** [comment]
@rfairley as far as I remember Cockpit symlinks it, too, doesn't it? I hit the same issue with greenboot, though. Not sure what'll make it go away.

**Comment by rfairley** [comment]
@LorbusChris yes, I believe so. Just been looking at this https://github.com/cockpit-project/cockpit/blob/master/src/ws/Makefile-ws.am#L178, seems this is where the symlink gets created.

**Comment by dustymabe** [comment]
> Any thoughts on this - too verbose?

i probably wouldn't bother with showing available users (since you already logged in at this point). For an available update it might be nice to show the current version and the available version in the update with the corresponding short commit id

```
---- ostree
current: 28.1 (aabbccd)
available: 28.2 (eeffggh)
```

**Comment by rfairley** [comment]
@dustymabe this looks good, will add this in:

Some other items in the motd are a message saying if a deployment is pending, and security advisories if there are any. The login so far will look like:

```
Fedora (28 (Atomic Host))

---- rpm-ostree
Deployment pending on next reboot
Security advisories:
  3 unknown severity, 1 low, 1 moderate, 2 important

---- ostree
current: 28.1 (aabbccd)
available: 28.2 (eeffggh)

Failed Units: 1
  example.service
```

(with "failed units" coming from `/etc/profile.d/`, not motd, generated by https://github.com/rfairley/fedora-coreos-login-messages/blob/master/usr/share/coreos/coreos-profile.sh)

Feedback on this is appreciated!

**Comment by sgallagher** [comment]
> Can be tested by following: https://github.com/rfairley/fedora-coreos-login-messages#installation. This should show the issue file as expected after completing step 7.
> 
> One remaining problem is that the file symlinked as `/etc/motd.d/fedora-user-messages.motd -> /run/fedora-user-messages.motd` is not being displayed by PAM after SSH login. `cat /run/fedora-user-messages.motd` shows the file is being generated correctly by the systemd units. Copying the generated file so it is a regular file in`/etc/motd.d/` causes the motd to be shown after SSH login - this makes me think symlinks in `/etc/motd.d/` are not being followed by PAM. Looking into this to see if cockpit got this to work.

Check for SELinux denials. See https://bugzilla.redhat.com/show_bug.cgi?id=1584167

**Comment by sgallagher** [comment]
@dustymabe 
> right now the rpms are named `fedora-user-messages-*`. Which we might have to rebrand if this package ever got passed along to RHEL/centos. Should we just name it something generic? @sgallagher do you have a suggestion?

Beyond that, we really need to avoid having "Fedora" branding anywhere outside of the fedora-release, fedora-logos and fedora-repos packages for RHEL/CentOS and Fedora Remixes (which aren't allowed to use the Fedora brand). So yeah, please come up with a more generic name.


If you want this to show up for all Fedora installations, you'll need to propose this as a [Fedora Change](https://fedoraproject.org/wiki/Changes/Policy). In that case (if approved by FESCo), we'll want to merge this code into the fedora-release package.

**Comment by rfairley** [comment]
@sgallagher 
> Check for SELinux denials. See https://bugzilla.redhat.com/show_bug.cgi?id=1584167

That was it, thanks! Shows after `sudo chcon -t etc_t /run/fedora-user-messages.motd`

**Comment by ffooffer** [comment]
> When you log in to container linux you get a nice prompt that tells you some information about the machine.
> 
> ```
> $ ssh core@192.168.122.136
> Container Linux by CoreOS stable (1800.7.0)
> Failed Units: 1
>   example.service
> core@localhost ~ $ 
> ```
> On the serial console there is a nice message telling you what the IP address is for the machine.
> 
> ```
> This is localhost (Linux x86_64 4.14.16-coreos) 14:19:32
> SSH host key: SHA256:yP+/44/bfuj6UKHdUwAVURsO3y6haKLKfSFNcnmn7bY (ECDSA)
> SSH host key: SHA256:gGDZ/JQzwL76UpT29dyZ/M6Zua7QvGyegP8aTLc/D+Y (DSA)
> SSH host key: SHA256:nQEysCYP3hZgkus2+e28KQGrs0pRI2NOgJGQ6L8PnyU (RSA)
> SSH host key: SHA256:A3c6toZ3/eTMKNDmyyG9CYUSWsdSunmTeOC68iuDfAg (ED25519)
> eth0: 192.168.122.36 fe80::5054:ff:fe85:43a6
> ```
> we should carry these nice features forward for Fedora CoreOS.

[](url)

**Comment by dustymabe** [comment]
@ffooffer - did you mean to add a comment?

**Comment by rfairley** [comment]
@dustymabe yes - good to close this now with the flow of the messages coming from an RPM. We'll want to have `console-login-helper-messages` come from official repos eventually, but that will be tracked with COREOS-756 (`Have console-login-helper-messages accepted as a package in Fedora upstream`).
